peerManager.go

```diff
@@ -0,0 +1,256 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/cache/ttl"
+	core "github.com/libp2p/go-libp2p-core"
+	"github.com/libp2p/go-libp2p-core/peer"
+	discovery "github.com/libp2p/go-libp2p-discovery"
+	"github.com/multiformats/go-multiaddr"
+	"go.uber.org/zap"
+)
+
+var (
+	errNoDiscoverer = errors.New("discoverer isn't set")
+	errNoPeers      = errors.New("no peers are found")
+)
+
+const _connectRetry = 5
+
+type peerManagerOpt func(cfg *peerManager)
+
+func withMaxPeers(num int) peerManagerOpt {
+	return func(pm *peerManager) {
+		pm.maxPeers = num
+	}
+}
+
+func withBlacklistTolerance(num int) peerManagerOpt {
+	return func(pm *peerManager) {
+		pm.blacklistTolerance = num
+	}
+}
+
+func withBlacklistTimeout(t time.Duration) peerManagerOpt {
+	return func(pm *peerManager) {
+		pm.blacklistTimeout = t
+	}
+}
+
+func withAdvertiseInterval(t time.Duration) peerManagerOpt {
+	return func(pm *peerManager) {
+		pm.blacklistTimeout = t
+	}
+}
+
+type peerManager struct {
+	routing  *discovery.RoutingDiscovery
+	host     core.Host
+	ns       string
+	maxPeers int // unlimited peers when maxPeers = 0
+
+	bootstrap      map[core.PeerID]bool
+	blacklist      *ttl.Cache
+	advertiseQueue chan string
+
+	blacklistTolerance int
+	blacklistTimeout   time.Duration
+	advertiseInterval  time.Duration
+}
+
+func newPeerManager(host core.Host, routing *discovery.RoutingDiscovery, ns string, opts ...peerManagerOpt) *peerManager {
+	rand.Seed(time.Now().UnixNano())
+	pm := &peerManager{
+		host:              host,
+		routing:           routing,
+		ns:                ns,
+		bootstrap:         make(map[peer.ID]bool),
+		advertiseInterval: 3 * time.Hour, // DHT provider record is republished every 3hrs by default
+	}
+	for _, opt := range opts {
+		opt(pm)
+	}
+	if pm.blacklistTimeout > 0 {
+		bl, _ := ttl.NewCache(ttl.AutoExpireOption(pm.blacklistTimeout))
+		pm.blacklist = bl
+	} else {
+		bl, _ := ttl.NewCache()
+		pm.blacklist = bl
+	}
+
+	return pm
+}
+
+func (pm *peerManager) JoinOverlay() {
+	pm.advertiseQueue = make(chan string)
+	ticker := time.NewTicker(pm.advertiseInterval)
+	go func(pm *peerManager) {
+		for {
+			select {
+			case ns := <-pm.advertiseQueue:
+				_, err := pm.routing.Advertise(context.Background(), ns)
+				if err != nil {
+					Logger().Error("error when advertising.", zap.Error(err))
+				}
+			case <-ticker.C:
+				_, err := pm.routing.Advertise(context.Background(), pm.ns)
+				if err != nil {
+					Logger().Error("error when advertising.", zap.Error(err))
+				}
+			}
+		}
+	}(pm)
+}
+
+func (pm *peerManager) Advertise() error {
```

Comment:
also we don't need this? in `JoinOverlay()`, `ticker.C` will fire every 3 hours to keep advertising itself, right? maybe we can reduce to interval to like 30 min

Reply:
this is the func triggerred externally